tslib/src/caching/redis-service.ts

```diff
@@ -158,22 +149,67 @@ export class RedisService implements IRedisService {
     if (keys.length === 0) {
       return
     }
-    const pipeline = this.redis.pipeline()
+    const redis = await this.getRedis()
+    const pipeline = redis.pipeline()
     keys.forEach(c => {
       pipeline.del(c)
     })
     return pipeline.exec()
   }
 
-  async quit (): Promise<string> {
-    return this.redis.quit()
-  }
-
   async ttl (key: string): Promise<number | null> {
-    return this.redis.ttl(key)
+    const redis = await this.getRedis()
+    return redis.ttl(key)
   }
 
   async expire (key: string, ttl: number): Promise<any> {
-    return this.redis.expire(key, ttl)
+    const redis = await this.getRedis()
+    return redis.expire(key, ttl)
+  }
+}
+
+class RedisSingleton {
+  private static redisService: Redis.Redis
+
+  private static async getRemoteIp (): Promise<any> {
+    const requestUrl = config.RemoteIpCheckUrl
+    if (requestUrl === undefined) return 'remote url not configured'
+    try {
+      const requestConfig: AxiosRequestConfig = {
+        method: 'GET',
+        url: requestUrl
+      }
+      const response = await axios(requestConfig)
+      return response.data
+    } catch (error) {
+      console.error(`RedisSingleton.getRemoteIp: failed to make request to ${requestUrl} on function startup`)
+    }
+  }
+
+  private constructor () {}
+
+  private static readonly options: RedisOptions = {
+    port: Number(config.Redis.Port),
+    host: config.Redis.Host,
+    password: config.Redis.Key,
+    lazyConnect: true,
+    tls: config.Redis.useTLS ? { host: config.Redis.Host } : undefined
+  }
+
+  public static async getRedisService (): Promise<Redis.Redis> {
+    if (this.redisService === undefined) {
+      this.redisService = new Redis(this.options)
+      try {
+        console.log('RedisSingleton: attempting to connect to redis for first time...')
```

Comment:
wondering if we should leave this in, or at least narrate the connectivity sequence.  Just in case we need the visibility in the next few weeks 🤔  what do you think?

Reply:
Yes, okay, I can see the benefit in that.  If we log at 'info' level we will get a count of the new connections in a given time period.  We can then move the log-level up if we are comfortable with it.